### ISSUE TITLE
TCARouter (and all children) re-render too frequently

### ISSUE THREAD

**Issue by dannyhertz** [johnpatrickmorgan/TCACoordinators#18]
Am experiencing some strange SwiftUI bugs involving a view that is inside of a TCACoordinator and it seems like subview are re-rendering whenever any piece of state in the entire routes array changes. I believe the issue stems from this piece of code:

```
  public var body: some View {
    WithViewStore(store) { viewStore in <---------
      Router(
        viewStore.binding(
          get: routes,
          send: updateRoutes
        ),
        buildView: { screen, index in
          let id = identifier(screen, index)
          screenContent(scopedStore(id: id, screen: screen))
        }
      )
    }
  }
```

Since there is no scoping here in the `WithViewStore`, any change at all to the entire RoutesState causing the full View hierarchy to re-render. I'm trying to play around with different forms of removeDuplicates but am having trouble getting things to work. Will keep you posted but would love to know if you are aware of this issue.

Thanks again!

**Comment by dannyhertz** [comment]
I think I might be onto something. Seems like the `scopedStore` function was ignoring the updated state parameter and just returning the updated state from the surrounding WithViewStore, that "works" because WithViewStore updates so frequently (for every child change). I think my updated way (validated a bit in my own apps) causes *far* fewer re-renders and should give the same desired state. Just need to work a bit more on the removeDuplicates function to make it smarter (than just route count) about knowing when nav has changed. Will keep the thread posted!

```
  public var body: some View {
    WithViewStore(store, removeDuplicates: { routes($0).count == routes($1).count }) { viewStore in
      Router(
        viewStore.binding(
          get: routes,
          send: updateRoutes
        ),
        buildView: { screen, index in
          let id = identifier(screen, index)
          screenContent(
            store.scope(state: { routes($0)[index].screen }, action: { action(id, $0) })
          )
        }
      )
    }
  }
```

**Comment by johnpatrickmorgan** [comment]
Hi @dannyhertz, thanks for raising this issue, and investigating it! I completely agree with your rewrite of the `scopedStore` function. Coincidentally, I came to the same conclusion while looking into #19. 

I initially assumed your proposed `removeDuplicates` addition would not work, since it ignores changes to the individual screen states themselves, but I guess I was thinking about it wrongly, as it seems to work well as you say. I suppose the screen content _can_ safely be ignored because the `buildView` closure will take care of any changes to screen state?

I think the simplest way to incorporate the route style into that check would be like this:

```swift
WithViewStore(store, removeDuplicates: { routes($0).map(\.style) == routes($1).map(\.style) })
```

**Comment by dannyhertz** [comment]
Exactly, once the view is properly built the custom views can take over and observe what is needed. The container only need to re-render for actual navigation specific change. And yes, mapping on style is a great improvement! Cool if I throw this together in a PR to check out?

**Comment by johnpatrickmorgan** [comment]
Thanks @dannyhertz! I'd be very grateful for a PR if you have a chance. I think it would be good to include some sort of [bounds check](https://github.com/johnpatrickmorgan/TCACoordinators/commit/c87fd38c0a77b2d2785490b8fb0d9a4f79019f11) too, as I hit an out of bounds error when testing.

**Comment by johnpatrickmorgan** [comment]
This has now been resolved by changes in #20  and shipped in [v1.0.2](https://github.com/johnpatrickmorgan/TCACoordinators/releases/tag/0.1.2). Thanks @dannyhertz!
